### PR TITLE
fix: preserve push messages in cache until they are seen by client

### LINKED ISSUE #15205: Vaadin 23 Resynchronizing UI by client's request - If you are using push with a proxy, make sure the push timeout smaller proxy connection timeout
### Description of the bug

Mostly on my iPhone I still receive the following error  and my applications shows infinitive bold blue progress bar at the top of the page:

    c.v.f.s.communication.ServerRpcHandler   : Resynchronizing UI by client's request. A network message was lost before reaching the client and the client is reloading the full UI state. This typically happens because of a bad network connection with packet loss or because of some part of the network infrastructure (load balancer, proxy) terminating a push (websocket or long-polling) connection. If you are using push with a proxy, make sure the push timeout is set to be smaller than the proxy connection timeout

I understand that this issue maybe related to flaky mobile internet connection, but unfortunately it will be really hard to explain the same to every single user of my application. The application should be able to restore from such situation.

I use NGINX as a proxy and Spring Boot (Tomcat or Undertow) with Vaadin application.

In Vaadin application I use: 

    @Push(transport = Transport.WEBSOCKET_XHR)

My configs:

SpringBoot application.properties:

```
server.servlet.session.timeout=30m
server.servlet.session.tracking-modes = cookie
server.servlet.session.cookie.http-only=true
server.servlet.session.cookie.secure=true
server.servlet.session.cookie.max-age=365d

vaadin.heartbeatInterval=45
vaadin.pushLongPollingSuspendTimeout=180000
vaadin.maxMessageSuspendTimeout=5000
vaadin.compatibilityMode = false
```

NGINX:

proxy_http_version 1.1;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "Upgrade";
proxy_read_timeout     300;
proxy_connect_timeout  300;
proxy_send_timeout     300;

Atmosphere:

```
@Component
public class AtmosphereConfiguration implements ServletContextInitializer {

    @Override
    public void onStartup(ServletContext servletContext) {
        servletContext.setInitParameter("org.atmosphere.cpr.AtmosphereConfig.getInitParameter", "true");
        servletContext.setInitParameter("org.atmosphere.cpr.AtmosphereInterceptor", "org.atmosphere.interceptor.HeartbeatInterceptor");
        servletContext.setInitParameter("org.atmosphere.interceptor.HeartbeatInterceptor.paddingChar", "{\"command\":\"serverInfo\", \"data\":{}}");
        servletContext.setInitParameter("org.atmosphere.interceptor.HeartbeatInterceptor.heartbeatFrequencyInSeconds", "45");
    }

}
```

Related SO questions:

https://stackoverflow.com/questions/74350932/vaadin-23-resynchronizing-ui-by-clients-request-if-you-are-using-push-with-a
https://stackoverflow.com/questions/74418180/vaadin-23-websockets-iphone-updates-with-delays

### Expected behavior

Application should not freeze and continue to operate normally

### Minimal reproducible example

Try my application to use for some time on iPhone for example

### Versions

- Vaadin / Flow version: 23.2.8
- Java version: OpenJDK 64-Bit Server VM Corretto-17.0.1.12.1 (build 17.0.1+12-LTS, mixed mode, sharing)
- OS version: Debian
- Application Server : Tomcat or Undertow
- Proxy - NGINX

### PR DESCRIPTION
**Pull request by mcollovati**
## Description

Atmospehere caches messages when the client is disconnected, but unfortunately it may happen that a message does not reach the client because of network disconnection during async response write operation. In this case the message is not cached and will be completely lost, causing a UI resynchronization request.
This change preserves messages in broadcaster cache until the client confirms that they have been processed, by sending the last seen server sync identifier on reconnection.
This should prevent the need for a UI resynchronization. It may happen in some cases, e.g. back to online after being offline, that messages already seen will be sent to the client, but Flow will discard them.

Part of #15281
Fixes #15205 

## Type of change

- [X] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/server/communication/AtmospherePushConnection.java [hunk 1]
```diff
@@ -407,4 +410,33 @@ public static void enableAtmosphereDebugLogging() {
...
+
+    /**
+     * A {@link org.atmosphere.cpr.BroadcastFilter} that unwraps the message to
+     * be sent to the client from a {@link PushMessage} instance.
+     */
+    static final class PushMessageUnwrapFilter extends BroadcastFilterAdapter
+            implements Serializable {
+        @Override
+        public BroadcastAction filter(String broadcasterId,
+                AtmosphereResource r, Object originalMessage, Object message) {
```
[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/server/communication/AtmospherePushConnection.java:435]: @sonatype-lift ignore

File: flow-server/src/test/java/com/vaadin/flow/server/communication/LongPollingCacheFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    public void filter_messageNotYetSeen_addToCacheAndContinue() {
+        setTransport(AtmosphereResource.TRANSPORT.LONG_POLLING);
+        setSeenServerSyncIdHeader(2);
+        String broadcasterId = "broadcasterId";
+        BroadcastAction action = filter.filter(broadcasterId, resource,
+                originalMessage, message);
+        Assert.assertEquals("Expecting message not seen on client to be sent",
+                ACTION.CONTINUE, action.action());
+        Assert.assertSame(
+                "Message should not be altered by filter when aborting",
```
Comment (tepi) [inline flow-server/src/test/java/com/vaadin/flow/server/communication/LongPollingCacheFilterTest.java:124]: Should this be "Message should not be altered by filter when continuing"

File: flow-server/src/main/java/com/vaadin/flow/server/communication/LongPollingCacheFilter.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        return LoggerFactory.getLogger(LongPollingCacheFilter.class.getName());
+    }
+
+    @Override
+    public BroadcastAction filter(String broadcasterId, AtmosphereResource r,
+            Object originalMessage, Object message) {
+        if (originalMessage instanceof AtmospherePushConnection.PushMessage
+                && r.transport() == AtmosphereResource.TRANSPORT.LONG_POLLING
+                && r.getRequest().getHeader(SEEN_SERVER_SYNC_ID) != null) {
+            AtmospherePushConnection.PushMessage pushMessage = (AtmospherePushConnection.PushMessage) originalMessage;
```
[the remark this answers is not in the document]
Comment (tepi) [reply flow-server/src/main/java/com/vaadin/flow/server/communication/LongPollingCacheFilter.java:57]: @sonatype-lift ignore

**Review by tepi — APPROVED** [review]

**Comment by czp13** [comment]
Thanks for this fix, work @mcollovati, and review @tepi! 
I have a customer waiting for this to use it.  Any potential ETA when can this be merged back to master, and potentially released? 🙇

**Comment by mcollovati** [comment]
@czp13 the fix has been back ported and will soon be merged. It will be shipped with next Flow maintenance releases.

**Comment by czp13** [comment]
Thanks, @mcollovati!

**Comment by mcollovati** [comment]
@czp13 this fix has been released in Flow 23.3.4 and 23.2.10. Can you check if it solves issue with the customer project?

**Comment by czp13** [comment]
> @czp13 this fix has been released in Flow 23.3.4 and 23.2.10. Can you check if it solves the issue with the customer project?

Thank you @mcollovati. I meant I have a customer in a support ticket, I have no access to the customer code base, unfortunately, so no way to reproduce this. 
 
What I can and will do is that I notify the customer with the version containing the fix, and I get back to you and close my support ticket, if he confirms that your work solved the issue. Keep you in the loop!
(btw I think I will need an older Vaadin version, I will look for it).

**Comment by czp13** [comment]
History:
> @czp13 this fix has been released in Flow 23.3.4 and 23.2.10. Can you check if it solves the issue with the customer project?
> What I can and will do is that I notify the customer with the version containing the fix, and I get back to you and close my support ticket if he confirms that your work solved the issue. Keep you in the loop!

Notifying you here as well @mcollovati:
**The customer get back to me yesterday and notified me that they have their own implementation for the session expiration:** (so we cannot test it with them anymore unfortunately, anyway your fix is valuable and thank you for it 🙏 )

> We are happy to close this now. Unfortunately we won't be retesting on our side as we turned off the vaadin controlled session expiration and implemented our own version. It's great to know that it's fixed though.

I asked him to get back to us anytime they need our help regarding to this topic (or anything else).

**Pull request merged**